lib/src/repo.rs

```diff
@@ -1194,20 +1194,30 @@ impl MutableRepo {
 
     /// Order a set of commits in an order they should be rebased in. The result
     /// is in reverse order so the next value can be removed from the end.
-    fn order_commits_for_rebase(&self, to_visit: Vec<Commit>) -> BackendResult<Vec<Commit>> {
+    fn order_commits_for_rebase(
+        &self,
+        to_visit: Vec<Commit>,
+    ) -> BackendResult<Vec<(Commit, Vec<CommitId>)>> {
         let to_visit_set: HashSet<CommitId> =
             to_visit.iter().map(|commit| commit.id().clone()).collect();
         let mut visited = HashSet::new();
+        let mut parents_map = HashMap::new();
         // Calculate an order where we rebase parents first, but if the parents were
         // rewritten, make sure we rebase the rewritten parent first.
         let store = self.store();
-        dag_walk::topo_order_reverse_ok(
+        let ordered_commits = dag_walk::topo_order_reverse_ok(
             to_visit.into_iter().map(Ok),
             |commit| commit.id().clone(),
             |commit| -> Vec<BackendResult<Commit>> {
                 visited.insert(commit.id().clone());
                 let mut dependents = vec![];
-                for parent in commit.parents() {
+                let parent_ids = commit.parent_ids();
+                parents_map.insert(
+                    commit.id().clone(),
+                    parent_ids.iter().cloned().collect_vec(),
+                );
```

Comment:
Since we now have a static parents mapping, I don't think we need to build `(commit, parents)` pairs here. We can instead re-look up parents by `transform_descendants()`.